File: 2021/day1/day1.js
const fs = require('fs');
const path = require('path');

const part1 = input => {
  const measurements = input
    .trim()
    .split('\n')
    .map(s => parseInt(s, 10));

  let increased = 0;
  for (let index = 1; index < measurements.length; index++) {
    const cur = measurements[index];
    const prev = measurements[index - 1];
    if (cur > prev) increased++;
  }
  return increased;
};

const part2 = input => {
  let measurements = input
    .trim()
    .split('\n')
    .map(s => parseInt(s, 10));

  let increased = 0;
  for (let index = 3; index < measurements.length; index++) {
    let sum = (i) => measurements[i] + measurements[i - 1] + measurements[i - 2];
    const cur = sum(index);
    const prev = sum(index - 1);
    if (cur > prev) increased++;
  }
  return increased;
};

const input = fs.readFileSync(path.resolve(__dirname, './input.txt'), 'utf8');
console.log('day1 part1:', part1(input));
console.log('day1 part2:', part2(input));

File: 2021/day1/day1.clj
(require '[clojure.string :as str])

(def input
  (->>
   (slurp "2021/day1/input.txt")
   (str/split-lines)
   (map #(Integer/parseInt (str/trim %)))))

#_(def input '[199 200 208 210 200 207 240 269 260 263])

(defn part1 [xs]
  (->> xs
       (partition 2 1)
       (filter #(apply < %))
       count))

(defn part2 [xs]
  (->> xs
       (partition 3 1)
       (map #(apply + %))
       part1))

(println "day1 part1:" (part1 input))
(println "day1 part2:" (part2 input))
